include/albatross/src/linalg/block_diagonal.hpp

```diff
@@ -102,6 +126,55 @@ inline Eigen::Matrix<_Scalar, _Rows, _Cols> BlockDiagonalLDLT::sqrt_solve(
   return output;
 }
 
+template <class _Scalar, int _Options, typename _StorageIndex>
+inline Eigen::Matrix<_Scalar, Eigen::Dynamic, Eigen::Dynamic>
+BlockDiagonalLDLT::solve(
+    const Eigen::SparseMatrix<_Scalar, _Options, _StorageIndex> &rhs) const {
+  ALBATROSS_ASSERT(cols() == rhs.rows());
+  Eigen::Index i = 0;
+  Eigen::Matrix<_Scalar, Eigen::Dynamic, Eigen::Dynamic> output(rows(),
+                                                                rhs.cols());
+  for (const auto &b : blocks) {
+    const auto rhs_chunk = rhs.block(i, 0, b.rows(), rhs.cols());
+    output.block(i, 0, b.rows(), rhs.cols()) = b.solve(rhs_chunk);
+    i += b.rows();
+  }
+  return output;
+}
+
+template <typename Op, typename LhsType, typename RhsType>
+inline Eigen::MatrixXd BlockDiagonalLDLT::solve(
+    const Eigen::CwiseBinaryOp<Op, LhsType, RhsType> &rhs_orig) const {
+  ALBATROSS_ASSERT(cols() == rhs_orig.rows());
+  Eigen::MatrixXd rhs{rhs_orig};
+  return solve(rhs);
```

Comment:
what happens if you do; `solve(rhs_orig.eval())`?

Reply:
I can test it, but I don't think forcing an expression type is enough here.